eyeshade/workers/snapshots.js

```diff
@@ -0,0 +1,147 @@
+
+exports.initialize = async (debug, runtime) => {
+  await runtime.queue.create('update-snapshot-accounts')
+}
+
+// select * from account_balances
+const getAllAccountBalancesBefore = `
+select
+  account_transactions.account_type as "accountType",
+  account_transactions.account_id as "accountId",
+  coalesce(sum(account_transactions.amount), 0.0) as "balance"
+from account_transactions
+  where created_at < $1
+group by (account_transactions.account_type, account_transactions.account_id);
+`
+
+const getAllAccountBalances = `
+select * from account_balances
+`
+
+const upsertBalanceSnapshotAccounts = `
+insert into balance_snapshot_accounts (id, created_at, snapshot_id, account_id, account_type, balance)
+values ($1, $2, $3, $4, $5, $6)
+on conflict (id)
+do update
+  set balance = $6
+returning *;
+`
+
+const updateBalanceSnapshotWithTotals = `
+update balance_snapshots
+  set
+    completed = $2,
+    updated_at = $3
+where id = $1
+returning *;
+`
+
+exports.createDailySnapshot = createDailySnapshot
+exports.updateSnapshotAccounts = updateSnapshotAccounts
+
+exports.initialize = async (debug, runtime) => {
+  if ((typeof process.env.DYNO !== 'undefined') && (process.env.DYNO !== 'worker.1')) return
+
+  daily(debug, runtime)
+}
+
+function daily (debug, runtime) {
+  const interval = cron.parseExpression('* 0 * * *', {
+    utc: true
+  })
+  const date = interval.next().getTime()
+  setTimeout(() => createDailySnapshot(debug, runtime, date, daily), date - _.now())
+}
+
+async function createDailySnapshot (debug, runtime, date, next) {
+  const until = justDate(new Date(date))
+  const snapshotId = uuidV5(until, 'dc0befa2-37a4-4235-a5ce-dfc7d5408a78').toLowerCase()
+  const result = await updateSnapshotAccounts(debug, runtime, {
+    snapshotId,
+    until
+  })
+  next && next(debug, runtime)
+  return result
+}
+
+exports.workers = {
+  /* sent by POST /v1/snapshots/
+    { queue        : 'update-snapshot-accounts'
+    , message      :
+      { snapshotId : 'uuid',
+        until      : '2020-01-01T00:00:00.000Z'
+      }
+    }
+  */
+  'update-snapshot-accounts': updateSnapshotAccounts
+}
+
+function setupQuery (until = 'current') {
+  let query = getAllAccountBalances
+  const args = []
+  const d = new Date(until || new Date())
+  if (until !== 'current' && +d) {
+    query = getAllAccountBalancesBefore
+    args.push(d.toISOString())
+  }
+  return {
+    query,
+    args
+  }
+}
+
+async function updateSnapshotAccounts (debug, runtime, payload) {
+  const {
+    snapshotId,
+    until
+  } = payload
+  const client = await runtime.postgres.connect()
+  const roClient = await runtime.postgres.connect()
+  const {
+    query,
+    args
+  } = setupQuery(until)
+  let total = new BigNumber(0)
+  let count = new BigNumber(0)
+  const { rows } = await roClient.query(query, args)
+  roClient.release()
+  try {
+    const now = new Date()
+    await client.query('BEGIN')
+    for (let i = 0; i < rows.length; i += 1) {
```

Comment:
I think we can avoid this loop and use only one query for both the select and insert by doing something like
```
insert into balance_snapshot_accounts (account_id, account_type, balance)
select
  account_transactions.account_type as "accountType",
  account_transactions.account_id as "accountId",
  coalesce(sum(account_transactions.amount), 0.0) as "balance"
from account_transactions
  where created_at < $1
group by (account_transactions.account_type, account_transactions.account_id);
```
(query is incomplete)

https://www.postgresqltutorial.com/postgresql-insert/